src/rope.ml

```diff
@@ -0,0 +1,86 @@
+type t =
+  | Str of Cstruct.t * int * int
+  | App of t * t * int * int
+
+let length = function
+  | Str (_, _, len)
+  | App (_, _, len, _) -> len
+
+let empty = Str (Cstruct.empty, 0, 0)
+
+let height = function
+  | Str _ -> 0
+  | App (_, _, _, h) -> h
+
+let append = function
+  | Str (_,_,0), t | t, Str (_,_,0) ->  t
+  | t1, t2 -> 
+      App (t1, t2, length t1 + length t2, 1 + Int.max (height t1) (height t2))
+
+
+let rec sub t start stop =
+  if start == 0 && stop = length t
+  then t else match t with
+    | Str (str, off, _) ->
+        Str (str, off + start, stop - start)
+    | App (l, r, _, _) ->
+        let len = length l in
+        if stop <= len then sub l start stop
+        else if start >= len then sub r (start - len) (stop - len)
+        else append (sub l start len, sub r 0 (stop - len))
+
+let sub t ~off ~len =
+  let stop = off + len in
```

Comment:
I've a bad feeling about the potential integer overflow... unlikely to happen in practice, but I'm undecided whether we should check that!?

Reply:
If we check the call of `Rope.sub`, it is only called with `off:0`, so if an overflow appears, it must be elsewhere this addition.

Reply:
I fixed it with my last commit :+1:.